Normalizing config paths

JLSJamesShi opened the pull request:
The relative path will now be concatenated with the base path in a way such that the parent and current directory symbols are removed from the resulting path.

Test for normalizing an absolute path has been added.
Test for normalizing a concatenation of an absolute path and a relative path has been added.

JLSJamesShi committed:
paths are now normalized

JLSJamesShi committed:
Tests added for path resolution

sunjay commented on src/config.rs:
```diff
@@ -12,13 +14,33 @@ pub struct UnresolvedPath(std::path::PathBuf);
             base_dir.join(path)
+        };
+
+        let mut components = path.components().peekable();
```
Can you add a link (in a comment) to the place we lifted this code from. Will be useful in the future if we need to debug.

sunjay replied:
Also add a comment that describes that while we have an absolute path now, we need to normalize it to remove "." and ".." because Windows does not support that.

sunjay replied:
It looks like we don't actually use the fact that this is peekable? I think you are probably good to just remove the components variable entirely and move its contents to the for loop. Maybe we're missing something from the original code that needed this? Did you copy the entire algorithm?

JLSJamesShi replied:
Comment added + included the peek() section of the code.

sunjay commented on src/config.rs:
```diff
@@ -3,6 +3,8 @@ use std::num::{NonZeroUsize, NonZeroU32};
 use vek::{Vec3, Rgba};
 use serde::{Serialize, Deserialize};
 
+use std::path::{Path, Component};
```
Add a comment here that says that we are explicitly choosing not to import PathBuf because we don't want anyone in this module using it. We want every path in this module to be an UnresolvedPath.

JLSJamesShi replied:
Comment added.

sunjay commented on src/config.rs:
```diff
@@ -12,13 +14,33 @@ pub struct UnresolvedPath(std::path::PathBuf);
+        };
+
+        let mut components = path.components().peekable();
+        let mut ret = std::path::PathBuf::new();
```
Let's call this "normalized_path" to be more descriptive. I know the original code didn't choose good variable names, but that doesn't mean we have to follow their lead ;)

JLSJamesShi replied:
Variable renamed.

sunjay commented on src/config.rs:
```diff
@@ -12,13 +14,33 @@ pub struct UnresolvedPath(std::path::PathBuf);
+        for component in components {
+            match component {
+                Component::Prefix(..) => unreachable!(),
+                Component::RootDir => {
```
Let's remove the curly braces and make this more compact. No need to break out to the next line when there's just a single short expression.

JLSJamesShi replied:
Curly braces removed for RootDir.
The one for ParentDir is unchanged as I had to keep it a statement, since pop returns a bool.

sunjay commented on src/config.rs:
```diff
@@ -184,6 +206,28 @@ fn default_background() -> Rgba<f32> { Rgba {r: 0.0, g: 0.0, b: 0.0, a: 0.0} }
 
+    #[test]
+    fn resolve_absolute_path() {
+        let base_path = "/spritec/sample/bigboi";
```
Let's add a fake "/home/yourname/" prefix to all of these so no one thinks we actually expect spritec to be at the root of the file system. Also, we definitely need some tests with Windows style paths (including backslashes!) since this was originally a Windows bug.

JLSJamesShi replied:
Test path names were updated.
Relative path test for Windows was added.

sunjay commented on src/config.rs:
```diff
@@ -184,6 +206,28 @@ fn default_background() -> Rgba<f32> { Rgba {r: 0.0, g: 0.0, b: 0.0, a: 0.0} }
+        resolve_check(base_path, input_path, output_path);
+    }
+
+    fn resolve_check(base: &str, input: &str, output: &str) {
```
If you're curious, I can show you how we can use a Rust macro instead of this function to improve the tests and improve the output when tests fail. 

It's only a slight quality of life improvement, so it's definitely not necessary, but it might be fun to see if you're curious. :)

JLSJamesShi replied:
I have been taught by the master

sunjay commented on src/config.rs:
```diff
@@ -184,6 +206,28 @@ fn default_background() -> Rgba<f32> { Rgba {r: 0.0, g: 0.0, b: 0.0, a: 0.0} }
+    #[test]
+    fn resolve_relative_path() {
+        let base_path = "/spritec/sample/bigboi";
+        let input_path = "../../src/bin";
```
I would like to see a test where there are more ".." than there are parent directories and also one where you mix both ".." and "." syntax (normal usage and some weird combinations you wouldn't normally see)

(You should in general always try to consider what the edge/error cases are and add tests for those too)

sunjay reviewed (changes requested):
This is awesome! Very nicely done James! 

The vast majority of my comments are for small nitpicks and documentation things. Other than that there are a few for a couple of things I think were missed that I've added comments for as well. 

Happy to merge this once everything is resolved. Please don't hesitate to ask me any questions about any of my feedback. 

Great work! :tada:

sunjay commented:
Please rebase from master when you get the chance. Just merged #52.

JLSJamesShi committed:
Fixed path resolutions for Window's paths and added more resolution test cases

JLSJamesShi committed:
fixed some of the issues brought up in the PR

JLSJamesShi committed:
added path prefix matching (used for windows paths)

JLSJamesShi commented:
Could someone running Windows please cargo test this?

sunjay commented:
Can you add windows to our `.travis.yml` file? From [this blog post](https://blog.travis-ci.com/2018-10-11-windows-early-release) it seems like it might be as easy as adding this to that file near the top somewhere:

```yml
os:
  - windows
  - linux
  - osx
```

If it ends up not working for whatever reason and it's too complicated to fix this can be a future PR.

In the meantime, you may want to request that @daose run `cargo test` on his windows machine.

JLSJamesShi committed:
updated travis file to test for different operating systems

JLSJamesShi committed:
moved os lines from bottom of travis file to the top

daose commented:
fyi the tests all pass! but unfortunately the whole thing doesn't work 😢 

```
PS C:\Users\STUDENT\dev\spritec> cargo run .\samples\bigboi\spritec.toml
   Compiling spritec v0.1.0 (C:\Users\STUDENT\dev\spritec)
    Finished dev [optimized + debuginfo] target(s) in 38.12s
     Running `target\debug\spritec.exe .\samples\bigboi\spritec.toml`
tobj::load_obj - failed to open "C:Users\\STUDENT\\dev\\spritec\\samples\\bigboi\\obj\\bigboi_rigged_000005.obj" due to The system cannot find the path specified. (os error 3)
tobj::load_obj - failed to open "C:Users\\STUDENT\\dev\\spritec\\samples\\bigboi\\obj\\bigboi_rigged_000001.obj" due to The system cannot find the path specified. (os error 3)
tobj::load_obj - failed to open "C:Users\\STUDENT\\dev\\spritec\\samples\\bigboi\\obj\\bigboi_rigged_000001.obj" due to The system cannot find the path specified. (os error 3)
tobj::load_obj - failed to open "C:Users\\STUDENT\\dev\\spritec\\samples\\bigboi\\obj\\bigboi_rigged_000001.obj" due to The system cannot find the path specified. (os error 3)
tobj::load_obj - failed to open "C:Users\\STUDENT\\dev\\spritec\\samples\\bigboi\\obj\\bigboi_rigged_000005.obj" due to The system cannot find the path specified. (os error 3)
tobj::load_obj - failed to open "C:Users\\STUDENT\\dev\\spritec\\samples\\bigboi\\obj\\bigboi_rigged_000001.obj" due to The system cannot find the path specified. (os error 3)
tobj::load_obj - failed to open "C:Users\\STUDENT\\dev\\spritec\\samples\\bigboi\\obj\\bigboi_rigged_000005.obj" due to The system cannot find the path specified. (os error 3)
tobj::load_obj - failed to open "C:Users\\STUDENT\\dev\\spritec\\samples\\bigboi\\obj\\bigboi_rigged_000005.obj" due to The system cannot find the path specified. (os error 3)
Error: GltfError(Io(Os { code: 3, kind: NotFound, message: "The system cannot find the path specified." }))
error: process didn't exit successfully: `target\debug\spritec.exe .\samples\bigboi\spritec.toml` (exit code: 1)
PS C:\Users\STUDENT\dev\spritec>
```

I think it's missing the separator right after `C:`

sunjay commented:
Thanks for testing that Bill! James when you fix this make sure you add a unit test (a regression test) so we'll know if it ever doesn't work in the future.

JLSJamesShi committed:
Added support for resolving verbatim paths

sunjay committed:
Removed .travis.yml file that should no longer exist

sunjay committed:
Code review fixes

- importing PathBuf in resolve is fine
- simplified code in resolve
- reorganized tests to clearly show that all relevant cases are checked
  on each platform
- minor formatting tweaks

sunjay reviewed (approved):
Great work @JLSJamesShi! Assuming the build passes, this can be merged.

We'll go over the code review changes tomorrow.

:tada: :tada: :tada:
